tests/trace_server/run_as_user/cross_process_trace_server.py

```diff
@@ -0,0 +1,616 @@
+WORKER_LOOP_TIMEOUT_SECONDS = 1.0
+WORKER_SHUTDOWN_TIMEOUT_SECONDS = 5.0
+STREAM_END_SIGNAL = "STREAM_END"
+STOP_SIGNAL = "STOP"
+
+
+# Protocol for trace server methods to improve type safety
+class TraceServerMethod(Protocol):
+    """Protocol for trace server methods that take a BaseModel and return a BaseModel."""
+
+    def __call__(self, request: BaseModel) -> BaseModel: ...
+
+
+class StreamingTraceServerMethod(Protocol):
+    """Protocol for streaming trace server methods that return an Iterator."""
+
+    def __call__(self, request: BaseModel) -> Iterator[Any]: ...
+
+
+class EmptyPayload(BaseModel):
+    """Empty payload for control messages."""
+
+    pass
+
+
+class CrossProcessTraceServerError(Exception):
+    """Base exception for cross-process trace server errors."""
+
+    pass
+
+
+class RequestQueueItem(BaseModel):
+    """
+    Represents a request sent from child process to main process.
+
+    Attributes:
+        request_id: Unique identifier for tracking this specific request
+        method: Name of the TraceServerInterface method to invoke
+        payload: The request payload to pass to the method
+    """
+
+    request_id: str
+    method: str
+    payload: BaseModel
+
+
+class ResponseQueueItem(BaseModel):
+    """
+    Represents a response sent from main process to child process.
+
+    Attributes:
+        request_id: Matches the request_id from the original RequestQueueItem
+        error: Error message if the request failed, or "STREAM_END" for streaming termination
+        payload: The response payload from the method execution
+    """
+
+    request_id: str
+    error: Optional[str] = None
+    payload: Optional[BaseModel] = None
+
+
+class CrossProcessTraceServerSender(tsi.TraceServerInterface):
+    """
+    TraceServerInterface implementation that sends requests to another process via queues.
+
+    This class runs in the child process and communicates with a receiver in the main process.
+    It handles:
+    - Unique request ID generation to track responses
+    - Out-of-order response handling (responses may not arrive in request order)
+    - Streaming method support with proper termination signaling
+    - Timeout handling for stuck requests
+
+    The sender maintains a pending requests cache to handle out-of-order responses,
+    which can occur when multiple requests are in flight simultaneously.
+    """
+
+    def __init__(
+        self,
+        request_queue: multiprocessing.Queue,
+        response_queue: multiprocessing.Queue,
+        lock: multiprocessing.Lock,
+    ):
+        self.request_queue = request_queue
+        self.response_queue = response_queue
+        self._request_id_counter = 0
+        # Cache for responses that arrive before we're waiting for them
+        # TODO: Add periodic cleanup of old entries to prevent memory leaks
+        # TODO: Consider adding maximum cache size with LRU eviction
+        self._pending_requests: dict[str, ResponseQueueItem] = {}
```

Comment:
if we are worried about this, maybe there would be a convenient place to log to datadog the state of the queues? we can log or attach it to a span? might help catch errors before they blow up